File: scripts/sku_registry.py
#!/usr/bin/env python3
"""Stable SKU registry (Python mirror of sku_registry.mjs)."""

from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
REGISTRY_PATH = ROOT / "data" / "sku_registry.json"
PRODUCTS_PATH = ROOT / "public" / "products.json"


def norm(s: str) -> str:
    s = (s or "").lower()
    s = re.sub(r"\s+", " ", s)
    s = s.replace("«", "").replace("»", "").replace('"', "").replace("'", "")
    return s.strip()


def product_key(name: str, weight: str = "") -> str:
    return f"{norm(name)}|{norm(weight)}"


def load_registry() -> dict:
    try:
        data = json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))
    except Exception:
        data = {}
    data.setdefault("version", 1)
    data.setdefault("next_n", 1)
    data.setdefault("by_key", {})
    data.setdefault("retired", {})
    if not isinstance(data["next_n"], int) or data["next_n"] < 1:
        data["next_n"] = 1
    return data


def save_registry(reg: dict) -> None:
    REGISTRY_PATH.parent.mkdir(parents=True, exist_ok=True)
    REGISTRY_PATH.write_text(json.dumps(reg, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def bootstrap_from_products(reg: dict) -> dict:
    if reg.get("by_key"):
        return reg
    try:
        products = json.loads(PRODUCTS_PATH.read_text(encoding="utf-8")).get("products") or []
    except Exception as e:
        print(f"  ⚠️  sku_registry bootstrap skipped: {e}")
        return reg
    max_n = 0
    for p in products:
        m = re.match(r"^KD-(\d+)$", str(p.get("sku") or ""), re.I)
        if not m:
            continue
        n = int(m.group(1))
        max_n = max(max_n, n)
        key = product_key(p.get("name", ""), p.get("weight", ""))
        if key not in reg["by_key"]:
            reg["by_key"][key] = str(p["sku"]).upper()
    reg["next_n"] = max(reg["next_n"], max_n + 1)
    print(
        f"  🔑 sku_registry bootstrapped from products.json: "
        f"{len(reg['by_key'])} keys, next=KD-{reg['next_n']:03d}"
    )
    return reg


def assign_sku(reg: dict, name: str, weight: str = "") -> str:
    key = product_key(name, weight)
    if key in reg["by_key"]:
        return reg["by_key"][key]
    used = set(reg["by_key"].values()) | set((reg.get("retired") or {}).keys())
    n = reg["next_n"]
    while True:
        sku = f"KD-{n:03d}"
        n += 1
        if sku not in used:
            break
    reg["by_key"][key] = sku
    reg["next_n"] = n
    return sku


def retire_missing(reg: dict, live_keys) -> int:
    live = set(live_keys)
    retired = 0
    for key, sku in list(reg["by_key"].items()):
        if key not in live:
            reg["retired"][sku] = key
            del reg["by_key"][key]
            retired += 1
    if retired:
        print(f"  🗄  sku_registry: retired {retired} SKU(s) (kept reserved, not reused)")
    return retired

File: scripts/sku_registry.mjs
/**
 * Stable SKU registry — product URLs must not reshuffle when Google Sheets
 * rows are deleted/reordered. Fingerprint = normalized name + weight.
 *
 * File: data/sku_registry.json
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, '..');
const REGISTRY_PATH = path.join(ROOT, 'data', 'sku_registry.json');
const PRODUCTS_PATH = path.join(ROOT, 'public', 'products.json');

function norm(s) {
  return String(s || '')
    .toLowerCase()
    .replace(/\s+/g, ' ')
    .replace(/[«»"'″]/g, '')
    .trim();
}

export function productKey(name, weight = '') {
  return `${norm(name)}|${norm(weight)}`;
}

function emptyRegistry() {
  return { version: 1, next_n: 1, by_key: {}, retired: {} };
}

export function loadRegistry() {
  try {
    const data = JSON.parse(fs.readFileSync(REGISTRY_PATH, 'utf8'));
    if (!data.by_key) data.by_key = {};
    if (!data.retired) data.retired = {};
    if (!Number.isInteger(data.next_n) || data.next_n < 1) data.next_n = 1;
    return data;
  } catch {
    return emptyRegistry();
  }
}

export function saveRegistry(reg) {
  fs.mkdirSync(path.dirname(REGISTRY_PATH), { recursive: true });
  fs.writeFileSync(REGISTRY_PATH, JSON.stringify(reg, null, 2) + '\n', 'utf8');
}

/** Bootstrap from current products.json so existing URLs stay put. */
export function bootstrapFromProducts(reg) {
  if (Object.keys(reg.by_key).length > 0) return reg;
  try {
    const data = JSON.parse(fs.readFileSync(PRODUCTS_PATH, 'utf8'));
    const products = data.products || [];
    let maxN = 0;
    for (const p of products) {
      const m = String(p.sku || '').match(/^KD-(\d+)$/i);
      if (!m) continue;
      const n = parseInt(m[1], 10);
      if (n > maxN) maxN = n;
      const key = productKey(p.name, p.weight);
      // Prefer first occupant if collision
      if (!reg.by_key[key]) reg.by_key[key] = p.sku.toUpperCase();
    }
    reg.next_n = Math.max(reg.next_n, maxN + 1);
    console.log(`  🔑 sku_registry bootstrapped from products.json: ${Object.keys(reg.by_key).length} keys, next=KD-${String(reg.next_n).padStart(3, '0')}`);
  } catch (e) {
    console.warn(`  ⚠️  sku_registry bootstrap skipped: ${e.message}`);
  }
  return reg;
}

/**
 * Pack size or a clarifying word in the name changed, same product.
 * Reuse the URL only when the match is unique — never steal a sibling SKU.
 */
function findMigratableEntry(reg, name, weight) {
  const n = norm(name);
  const w = norm(weight);

  const sameName = Object.entries(reg.by_key).filter(([k]) => k.split('|')[0] === n);
  if (sameName.length === 1) return sameName[0];

  const sameWeightCloseName = Object.entries(reg.by_key).filter(([k]) => {
    const [en, ew] = k.split('|');
    if (ew !== w || en === n) return false;
    return n.startsWith(`${en} `) || en.startsWith(`${n} `);
  });
  if (sameWeightCloseName.length === 1) return sameWeightCloseName[0];

  return null;
}

/**
 * Assign a stable SKU for this product. Reuses registry entry; never
 * reassigns an existing SKU to a different fingerprint.
 */
export function assignSku(reg, name, weight = '') {
  const key = productKey(name, weight);
  if (reg.by_key[key]) return reg.by_key[key];

  const migrated = findMigratableEntry(reg, name, weight);
  if (migrated) {
    const [oldKey, sku] = migrated;
    delete reg.by_key[oldKey];
    reg.by_key[key] = sku;
    console.log(`  🔑 sku_registry: ${sku} ${oldKey} → ${key}`);
    return sku;
  }

  // Find free KD-NNN starting at next_n (skip any still mapped)
  const used = new Set(Object.values(reg.by_key));
  for (const sku of Object.keys(reg.retired || {})) used.add(sku);

  let n = reg.next_n;
  let sku;
  do {
    sku = `KD-${String(n).padStart(3, '0')}`;
    n += 1;
  } while (used.has(sku));

  reg.by_key[key] = sku;
  reg.next_n = n;
  return sku;
}

/** Mark SKUs absent from this sync as retired (do not reuse). */
export function retireMissing(reg, liveKeys) {
  const live = new Set(liveKeys);
  const dropped = [];
  for (const [key, sku] of Object.entries(reg.by_key)) {
    if (!live.has(key)) {
      dropped.push([key, sku]);
      delete reg.by_key[key];
    }
  }
  const stillUsed = new Set(Object.values(reg.by_key));
  let retired = 0;
  for (const [key, sku] of dropped) {
    if (stillUsed.has(sku)) continue;
    reg.retired[sku] = key;
    retired += 1;
  }
  if (retired) {
    console.log(`  🗄  sku_registry: retired ${retired} SKU(s) (kept reserved, not reused)`);
  }
  return retired;
}
